pkg/vmprovider/providers/vsphere/contentlibrary/content_library_cache.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package contentlibrary
+
+import (
+	"sync"
+	"time"
+
+	"github.com/vmware/govmomi/ovf"
+)
+
+type Cache struct {
+	mu      sync.Mutex
+	items   map[string]*itemCache
+	maxSize int
+}
+
+type itemCache struct {
+	ovfEnvelope     ovf.Envelope
+	contentVersion  string
+	lastUpdatedTime int64
+}
+
+// NewCache creates a new content library cache with the given item expiration.
+// It also starts a goroutine to clean up expired cache items at the given check interval.
+func NewCache(itemExpiration, expirationCheckInterval time.Duration, maxSize int) *Cache {
+	c := &Cache{items: make(map[string]*itemCache), maxSize: maxSize}
+
+	// The following go routine cleans up expired cache items at the given check interval.
+	go func() {
+		itemExpirationSec := int64(itemExpiration.Seconds())
+		for range time.Tick(expirationCheckInterval) {
+			c.mu.Lock()
+			for key, item := range c.items {
+				if time.Now().Unix()-item.lastUpdatedTime > itemExpirationSec {
+					log.V(5).Info("Deleting expired cache item", "itemID", key)
```

Comment:
Don't log here. Having a logger in a goroutine like this is a good way to hit an NPE when the system is shutting down or something else you didn't expect, this Goroutine is still running, and the logger is no longer initialized enough to work correctly. I'll show you a better way in a moment with the refactor I'm providing.

Reply:
Ah, I didn't realize that, thanks for pointing it out.

Reply:
Resolving this conversation as the `content_lirbary_cache` is replaced with the generic type cache.